packages/dds/sequence/docs/intervalCollectionQuerying.md

```diff
@@ -0,0 +1,240 @@
+		startPosition: number,
+	): IntervalCollectionIterator<TInterval>;
+	findOverlappingIntervals(startPosition: number, endPosition: number): TInterval[];
+	gatherIterationResults(
+		results: TInterval[],
+		iteratesForward: boolean,
+		start?: number,
+		end?: number,
+	): void;
+	getIntervalById(id: string): TInterval | undefined;
+	map(fn: (interval: TInterval) => void): void;
+	nextInterval(pos: number): TInterval | undefined;
+	previousInterval(pos: number): TInterval | undefined;
+	removeIntervalById(id: string): TInterval | undefined;
+}
+```
+
+This API splits roughly into operations and queries as follows:
+
+```typescript
+export class IntervalCollection<TInterval extends ISerializableInterval> extends TypedEventEmitter<
+	IIntervalCollectionEvent<TInterval>
+> {
+	add(start: number, end: number, intervalType: IntervalType, props?: PropertySet): TInterval;
+	addConflictResolver(conflictResolver: IntervalConflictResolver<TInterval>): void;
+	attachDeserializer(onDeserialize: DeserializeCallback): void;
+	change(id: string, start?: number, end?: number): TInterval | undefined;
+	changeProperties(id: string, props: PropertySet): void;
+	removeIntervalById(id: string): TInterval | undefined;
+}
+
+export class IntervalCollectionQueryMethods<TInterval extends ISerializableInterval> {
+	[Symbol.iterator](): IntervalCollectionIterator<TInterval>;
+	get attached(): boolean;
+	CreateBackwardIteratorWithEndPosition(
+		endPosition: number,
+	): IntervalCollectionIterator<TInterval>;
+	CreateBackwardIteratorWithStartPosition(
+		startPosition: number,
+	): IntervalCollectionIterator<TInterval>;
+	CreateForwardIteratorWithEndPosition(
+		endPosition: number,
+	): IntervalCollectionIterator<TInterval>;
+	CreateForwardIteratorWithStartPosition(
+		startPosition: number,
+	): IntervalCollectionIterator<TInterval>;
+	findOverlappingIntervals(startPosition: number, endPosition: number): TInterval[];
+	gatherIterationResults(
+		results: TInterval[],
+		iteratesForward: boolean,
+		start?: number,
+		end?: number,
+	): void;
+	getIntervalById(id: string): TInterval | undefined;
+	map(fn: (interval: TInterval) => void): void;
+	nextInterval(pos: number): TInterval | undefined;
+	previousInterval(pos: number): TInterval | undefined;
+}
+```
+
+The placement of "events" in this split doesn't matter too much, and they're arguably their own category.
+For practical purposes, a few of the APIs might be better left on interval collection--`attached`, `[Symbol.iterator]`, and `map`.
+
+## Splitting IntervalCollection into indexes
+
+Each query API above requires some particular indexing of the set of intervals to operate efficiently.
+[This commit](https://github.com/microsoft/FluidFramework/commit/01f20ee48198cd18011caa112f134c4ff337a8aa) shows an example factoring of `LocalReferenceCollection` into 3 parts.
+The key idea is the interval index interface:
+
+```typescript
+export interface IntervalIndex<TInterval extends ISerializableInterval> {
```

Comment:
is such as API necessary? couldn't the index use event + enumeration to build and bind itself, without the need to expose anything like this? Another option is to have an interface like this, but not export it, as consumers don't need it.

Overall, i think our strategy around internal is weak. Maybe someday we'll do literal d.ts pruning, so they aren't exposed, but even that will probably be at the interface/class level, not member. so i prefer to just not expose internal things where possible.

Reply:
Enumerate+eventing isn't enough unless we emit events before and after we change an interval due to comparator-on-mutable-objects issues.

If we don't expose "pick-and-choose which query APIs are supported", this interface doesn't need to be exposed to the package API though.

Reply:
I will say that even if we do want to go pick-and-choose or fully extensible, I like exposing this interface over some sort of before-change and after-change events. The semantics of the methods are clear. Marking them internal is probably not even the right call: the rest of this design works quite nicely with someone's custom indexing structure as long as it implements `add` and `remove`. These two methods just need to have documentation saying that user code never needs to call add or remove on their index for production scenarios, as it's handled for them by SharedString/IntervalCollection.

I agree with your points on internal. Another grievance is the fact that typetest tooling doesn't understand it, which means we open ourselves up to accidentally making real breaking changes whenever we refactor an internal API.